search/controllers/api/classic_parser.py

```diff
@@ -0,0 +1,190 @@
+"""
+Utility module for Classic API Query parsing.
+
+We use a recursive descent parser, implemented via :func:`parse_classic_query`.
+
+Tokens are themselves parsed using two helpers: :func:`_parse_operator` and
+:func:`_parse_field_query`.
+
+The final, parsed query is a :class:`domain.api.Phrase`, which is a nested
+set of Tuples::
+
+    >>> parse_classic_query("au:del_maestro AND ti:checkerboard")
+    ((Field.Author, 'del_maestro'), Operator.AND, (Field.Title, 'checkerboard'))
+
+See :module:`tests.test_classic_parser` for more examples.
+"""
+
+
+from typing import Any, List, Optional, Tuple, Union
+from operator import attrgetter
+
+from werkzeug.exceptions import BadRequest
+
+from ...domain.api import Phrase, Operator, Field, Term
+
+def parse_classic_query(query: str) -> Phrase:
+    """
+    Parse Classic API-style query string into app-native Phrase.
+
+    Iterates through each character in the string, applying a recursive-descent
+    parser and appropriately handling parens and quotes.
+
+    It iterates through each character:
+    1.  If a paren group is opened, we append to a list of start positions for
+        open paren groups.
+    2.  If a paren is closed, we pop out the most recent open paren from the 
+        stack of start positions and recursively parse the paren group, adding
+        it to our list of tokens.
+    3.  If a quote is opened, flip a quotation flag.
+    4.  If a space is encountered, append the token to the list of tokens and
+        reset the token start position.
+
+    Parameters
+    ----------
+    query : str
+        A Classic API query string.
+
+    Returns
+    -------
+    :class:`Phrase`
+        A tuple representing the query.
+
+    """
+    return _group_tokens(_cast_tokens(_tokenize_query_string(query)))
+
+
+def _tokenize_query_string(query: str) -> List[Union[str, Phrase]]:
+    """Tokenizes query string into list of strings and sub-Phrases."""
+    # Intializing state variables
+    tokens: List[Union[str, Phrase]] = []
+    token_start = 0
+    paren_group = []
+    in_quote = False
+
+    # Iterate through characters
+    for i, c in enumerate(query):
+        if c == '(':
+            paren_group.append(i) # add open paren start position to stack
```

Comment:
Always need two spaces before an inline comment. Also need capitalization and punctuation throughout.
```suggestion
            paren_group.append(i)  # Add open paren start position to stack.
```